File: go/modcdp/client/alias.go
package client

import (
	"encoding/json"
	"fmt"
	"strings"
)

type AliasSticky = map[string]any
type AliasJSONObject = map[string]any

type ModCDPAliasObject struct {
	Client *ModCDPClient
	Sticky AliasSticky
}

func NewModCDPAliasObject(client *ModCDPClient, sticky AliasSticky) ModCDPAliasObject {
	return ModCDPAliasObject{Client: client, Sticky: cloneAliasMap(sticky)}
}

func SendAliasCommand[T any](client *ModCDPClient, method string, params any, sticky AliasSticky, stickyParam string, stickyFields []string, unwrap string) (T, error) {
	stickyParams := []string{}
	if stickyParam != "" {
		stickyParams = []string{stickyParam}
	}
	return SendAliasCommandWithStickyParams[T](client, method, params, sticky, stickyParam, stickyParams, stickyFields, unwrap)
}

func SendAliasCommandWithStickyParams[T any](client *ModCDPClient, method string, params any, sticky AliasSticky, stickyParam string, stickyParams []string, stickyFields []string, unwrap string) (T, error) {
	var typed T
	if client == nil {
		return typed, fmt.Errorf("alias command %s requires a ModCDPClient", method)
	}
	rawParams, err := cdpParamsMap(params)
	if err != nil {
		return typed, err
	}
	mergeAliasStickyParams(rawParams, sticky, stickyParam, stickyParams, stickyFields)
	result, err := client.Send(method, rawParams)
	if err != nil {
		return typed, err
	}
	unwrapped, err := UnwrapAliasResult(result, unwrap)
	if err != nil {
		return typed, err
	}
	body, err := json.Marshal(unwrapped)
	if err != nil {
		return typed, err
	}
	if err := json.Unmarshal(body, &typed); err != nil {
		return typed, fmt.Errorf("%s result did not match typed alias result shape: %w", method, err)
	}
	return typed, nil
}

func OptionalAliasParams[T any](method string, params []T) (T, error) {
	var typed T
	if len(params) > 1 {
		return typed, fmt.Errorf("%s accepts at most one params object", method)
	}
	if len(params) == 1 {
		return params[0], nil
	}
	return typed, nil
}

func mergeAliasStickyParams(params map[string]any, sticky map[string]any, primaryStickyParam string, stickyParams []string, stickyFields []string) {
	if len(stickyParams) == 0 {
		mergeAliasSticky(params, sticky, primaryStickyParam, stickyFields)
		return
	}
	seen := map[string]bool{}
	for _, stickyParam := range stickyParams {
		if stickyParam == "" || seen[stickyParam] {
			continue
		}
		seen[stickyParam] = true
		if stickyParam != primaryStickyParam {
			if _, exists := params[stickyParam]; !exists {
				continue
			}
		}
		mergeAliasSticky(params, sticky, stickyParam, stickyFields)
	}
}

func AliasStickyFromResult(result any, unwrap string, stickyFields []string) (AliasSticky, error) {
	unwrapped, err := UnwrapAliasResult(result, unwrap)
	if err != nil {
		return nil, err
	}
	sticky, ok := unwrapped.(map[string]any)
	if !ok {
		body, err := json.Marshal(unwrapped)
		if err != nil {
			return nil, err
		}
		if err := json.Unmarshal(body, &sticky); err != nil {
			return nil, fmt.Errorf("alias sticky result did not decode to object: %w", err)
		}
	}
	if len(stickyFields) == 0 {
		return cloneAliasMap(sticky), nil
	}
	filtered := map[string]any{}
	for _, field := range stickyFields {
		if value, ok := sticky[field]; ok {
			filtered[field] = value
		}
	}
	return filtered, nil
}

func AliasArrayFromResult(result any, unwrap string) ([]any, error) {
	unwrapped, err := UnwrapAliasResult(result, unwrap)
	if err != nil {
		return nil, err
	}
	items, ok := unwrapped.([]any)
	if ok {
		return items, nil
	}
	body, err := json.Marshal(unwrapped)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(body, &items); err != nil {
		return nil, fmt.Errorf("alias result did not decode to array: %w", err)
	}
	return items, nil
}

func UnwrapAliasResult(result any, unwrap string) (any, error) {
	if unwrap == "" {
		return result, nil
	}
	current := result
	for _, part := range strings.Split(unwrap, ".") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		object, ok := current.(map[string]any)
		if !ok {
			body, err := json.Marshal(current)
			if err != nil {
				return nil, err
			}
			object = map[string]any{}
			if err := json.Unmarshal(body, &object); err != nil {
				return nil, fmt.Errorf("alias unwrap %q expected object at %q: %w", unwrap, part, err)
			}
		}
		value, ok := object[part]
		if !ok {
			return nil, fmt.Errorf("alias unwrap %q missing %q", unwrap, part)
		}
		current = value
	}
	return current, nil
}

func mergeAliasSticky(params map[string]any, sticky map[string]any, stickyParam string, stickyFields []string) {
	if params == nil || sticky == nil || len(sticky) == 0 {
		return
	}
	target := params
	if stickyParam != "" {
		existing, ok := params[stickyParam].(map[string]any)
		if !ok {
			existing = map[string]any{}
			if raw, exists := params[stickyParam]; exists && raw != nil {
				body, err := json.Marshal(raw)
				if err == nil {
					_ = json.Unmarshal(body, &existing)
				}
			}
			params[stickyParam] = existing
		}
		target = existing
	}
	fields := stickyFields
	if len(fields) == 0 {
		for key := range sticky {
			fields = append(fields, key)
		}
	}
	for _, field := range fields {
		if _, exists := target[field]; exists {
			continue
		}
		if value, ok := sticky[field]; ok {
			target[field] = value
		}
	}
}

func cloneAliasMap(source AliasSticky) AliasSticky {
	if source == nil {
		return AliasSticky{}
	}
	cloned := make(AliasSticky, len(source))
	for key, value := range source {
		cloned[key] = value
	}
	return cloned
}

File: python/modcdp/client/alias.py
from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from typing import Any, TypeVar

from pydantic import BaseModel

from .ModCDPClient import ModCDPClient

AliasSticky = dict[str, object]
AliasJSONObject = dict[str, object]
T = TypeVar("T")


class ModCDPAliasObject:
    def __init__(self, client: ModCDPClient, sticky: Mapping[str, object] | None = None) -> None:
        self.client = client
        self.sticky = _clone_alias_object(sticky or {})


def optional_alias_params(method: str, params: Sequence[T]) -> T | None:
    if len(params) > 1:
        raise ValueError(f"{method} accepts at most one params object")
    return params[0] if params else None


def send_alias_command_with_sticky_params(
    client: ModCDPClient,
    method: str,
    params: object,
    sticky: Mapping[str, object],
    sticky_param: str,
    sticky_params: Sequence[str],
    sticky_fields: Sequence[str],
    unwrap: str,
) -> object:
    raw_params = _params_to_alias_object(params)
    _merge_alias_sticky_params(raw_params, sticky, sticky_param, sticky_params, sticky_fields)
    return unwrap_alias_result(client._send_command(method, raw_params), unwrap)


def send_alias_command(
    client: ModCDPClient,
    method: str,
    params: object,
    sticky: Mapping[str, object],
    sticky_param: str,
    sticky_fields: Sequence[str],
    unwrap: str,
) -> object:
    return send_alias_command_with_sticky_params(
        client,
        method,
        params,
        sticky,
        sticky_param,
        [sticky_param] if sticky_param else [],
        sticky_fields,
        unwrap,
    )


def alias_sticky_from_result(result: object, unwrap: str, sticky_fields: Sequence[str]) -> AliasSticky:
    source = _params_to_alias_object(unwrap_alias_result(result, unwrap))
    if not sticky_fields:
        return _clone_alias_object(source)
    return {field: source[field] for field in sticky_fields if field in source}


def alias_array_from_result(result: object, unwrap: str) -> list[object]:
    unwrapped = unwrap_alias_result(result, unwrap)
    if isinstance(unwrapped, list):
        return unwrapped
    raise TypeError(f"alias unwrap {unwrap!r} expected array")


def unwrap_alias_result(result: object, unwrap: str) -> object:
    if not unwrap:
        return result
    current = result
    for part in unwrap.split("."):
        if not part:
            continue
        current_object = _params_to_alias_object(current)
        if part not in current_object:
            raise KeyError(f"alias unwrap {unwrap!r} missing {part!r}")
        current = current_object[part]
    return current


def _merge_alias_sticky_params(
    params: AliasJSONObject,
    sticky: Mapping[str, object],
    primary_sticky_param: str,
    sticky_params: Sequence[str],
    sticky_fields: Sequence[str],
) -> None:
    if not sticky_params:
        _merge_alias_sticky(params, sticky, primary_sticky_param, sticky_fields)
        return
    seen: set[str] = set()
    for sticky_param in sticky_params:
        if not sticky_param or sticky_param in seen:
            continue
        seen.add(sticky_param)
        if sticky_param != primary_sticky_param and sticky_param not in params:
            continue
        _merge_alias_sticky(params, sticky, sticky_param, sticky_fields)


def _merge_alias_sticky(
    params: AliasJSONObject,
    sticky: Mapping[str, object],
    sticky_param: str,
    sticky_fields: Sequence[str],
) -> None:
    if not sticky:
        return
    target = params
    if sticky_param:
        raw_target = params.get(sticky_param)
        target = dict(raw_target) if isinstance(raw_target, Mapping) else {}
        params[sticky_param] = target
    fields = sticky_fields or list(sticky.keys())
    for field in fields:
        if field in target:
            continue
        if field in sticky:
            target[field] = sticky[field]


def _params_to_alias_object(params: object) -> AliasJSONObject:
    if params is None:
        return {}
    if isinstance(params, BaseModel):
        return dict(params.model_dump(mode="json", exclude_none=True, by_alias=True))
    if isinstance(params, Mapping):
        return dict(params)
    return {}


def _clone_alias_object(source: Mapping[str, object]) -> AliasSticky:
    return json.loads(json.dumps(dict(source)))
